jaxopt/_src/lbfgs.py

```diff
@@ -53,7 +54,7 @@ def inv_hessian_product_leaf(v: jnp.ndarray,
   indices = (start + jnp.arange(history_size)) % history_size
 
   def body_right(r, i):
-    alpha = rho_history[i] * jnp.vdot(s_history[i], r)
+    alpha = rho_history[i] * jnp.vdot(s_history[i], r).real
```

Comment:
Just to be sure that I understand:
- in theory the dot product should be real because the function is real and one rounds it for numerical stability or does the algorithm really require taking the real part?

Reply:
The algorithm indeed requires to take the real part. Then it is identical to the situation when the complex arguments are split into real and imaginary parts. `jnp.vdot(x, y).real` could also be expressed as `jnp.vdot(x.real, y.real) + jnp.vdot(x.imag, y.imag)`, which might be more efficient, need to benchmark it. For the moment I chose the simpler expression.

Reply:
Great, we may add a function vdot_real in tree_utils like the one already present in jax (see https://github.com/google/jax/blob/1795b12a9f7bca118d75c981fadd61331680de57/jax/_src/scipy/sparse/linalg.py#L39). Using the second option (`jnp.vdot(x.real, y.real) + jnp.vdot(x.imag, y.imag)`) is a priori slightly faster I think.

Reply:
Some benchmarking, see
[https://gist.github.com/geggo/d1a990b7d7c6ce97180dbb00c15f1ca0]
shows that `jnp.vdot(x, y).real`  is actually faster than `jnp.vdot(x.real, y.real) + jnp.vdot(x.imag, y.imag)` when non jitted, no difference when jitted.

Furthermore, I checked that only if both arrays are complex, the second term of `jnp.vdot(x.real, y.real) + jnp.vdot(x.imag, y.imag)` is needed because it is zero if either x or y real arrays. It seems the jax implementation could be improved - or I am missing something?